### PR TITLE
Adding more support for interaction with lightning. Closes #2754

### LINKED ISSUE #2754: Lightning Bolt Spawning can't be intercepted by events
World#addWeatherEffect can have an event fire/cancel to mitigate this.

### PR DESCRIPTION
**Pull request by Deriboy**
Cause lightning spawns to trigger EntityJoinWorldEvent (#2754)
~~Add new LightningStrikeEvent that is fired when lightning strikes~~

There currently isn't any good way to interact with lightning, so many mods just spawn their own lightning. These changes allow mods to respond to naturally occurring lightning spawns as well as potentially canceling or moving it it.

### CODE REVIEW

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 1]
```diff
@@ -84,13 +84,7 @@
...
 import net.minecraftforge.event.brewing.PlayerBrewedPotionEvent;
 import net.minecraftforge.event.brewing.PotionBrewEvent;
-import net.minecraftforge.event.entity.EntityEvent;
-import net.minecraftforge.event.entity.EntityMobGriefingEvent;
-import net.minecraftforge.event.entity.EntityMountEvent;
-import net.minecraftforge.event.entity.EntityStruckByLightningEvent;
-import net.minecraftforge.event.entity.PlaySoundAtEntityEvent;
-import net.minecraftforge.event.entity.ProjectileImpactEvent;
-import net.minecraftforge.event.entity.ThrowableImpactEvent;
+import net.minecraftforge.event.entity.*;
```
Comment (LumberWizard) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: .*?

Comment (Deriboy) [reply src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Oh hey. Yeah I blame my IDE. I didn't notice that. I'll fix it.

**Comment by Deriboy** [comment]
I fixed the issue with imports.

**Review by liach — APPROVED** [review]

File: src/main/java/net/minecraftforge/event/entity/LightningStrikeEvent.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+package net.minecraftforge.event.entity;
+
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.ForgeEventFactory;
+import net.minecraftforge.fml.common.eventhandler.Cancelable;
+import net.minecraft.entity.effect.EntityLightningBolt;
+
+/**
+ * LightningStrikeEvent is fired when a lightning entity has struck a block.<br>
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/event/entity/LightningStrikeEvent.java]: Why do you have two different explanations as to when this event fires? Is it when it hits a block or an entity? Is it fired once per entity, or once per strike? Rephrase this whole thing to be clearer.

**Review by tterrag1098 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/event/entity/LightningStrikeEvent.java (hunk 2 above)
Comment (Deriboy) [reply src/main/java/net/minecraftforge/event/entity/LightningStrikeEvent.java]:
```suggestion
 * LightningStrikeEvent is fired once when a lightning entity strikes.<br>
```
What about this?

Comment (Deriboy) [reply src/main/java/net/minecraftforge/event/entity/LightningStrikeEvent.java]: @tterrag1098 I've updated the description to be a lot clearer.

**Comment by Deriboy** [comment]
I have discovered an odd issue with the current implementation of this. 
EntityJoinWorldEvent is supposed to be cancelable, however if you cancel a lightning bolt's EntityJoinWorldEvent (or change its position), fire will still spawn around its original position since EntityLightningBolt spawns fire *in its constructor*. This is obviously undesired, but I'm not sure what the best way to fix this would be. 
Ideally the fire creation code could be moved out of its constructor to a more appropriate location. 

Thoughts?

**Comment by Deriboy** [comment]
There are still changes requested on this PR, but I don't see any non-resolved requests. What still needs to be changed before this is production ready?

**Review by tterrag1098 — APPROVED** [review]

**Comment by tterrag1098** [comment]
This needs an update to resolve conflicts.

**Comment by Deriboy** [comment]
Merge conflicts fixed.

**Comment by Deriboy** [comment]
Merge conflicts fixed again.

**Review by mezz — APPROVED** [review]

**Comment by Deriboy** [comment]
As per some discussion on the discord a while back, the scope of this PR was reduced to only causing lightning spawns to trigger EntityJoinWorldEvent

**Pull request merged**